File: src/mvmctl/core/network/_lease_service.py
"""Network IP lease management.

This module handles IP address allocation, release, and lease tracking
for network management using a class-based design with network resolution.
"""

from __future__ import annotations

import ipaddress

from mvmctl.core.network._repository import LeaseRepository, NetworkRepository
from mvmctl.core.network._resolver import NetworkResolver
from mvmctl.exceptions import NetworkError
from mvmctl.models.network import NetworkItem, NetworkLeaseItem
from mvmctl.utils.network import NetworkUtils


class LeaseService:
    """Manages IP leases for a specific network.

    This class handles IP address allocation, release, and lease tracking
    for a single network identified by name or ID.

    Args:
        entity: Network name, ID prefix, or NetworkItem instance.
        repo: LeaseRepository instance for database operations.

    Raises:
        NetworkNotFoundError: If the network cannot be resolved.
    """

    def __init__(
        self, entity: str | NetworkItem, repo: LeaseRepository
    ) -> None:
        self._lease_repo = repo

        if isinstance(entity, NetworkItem):
            self._network = entity
        else:
            self._resolver = NetworkResolver(self._lease_repo)
            self._network = self._resolver.resolve(entity)

    @property
    def network_id(self) -> str:
        """Get the resolved network ID."""
        return self._network.id

    @property
    def network_name(self) -> str:
        """Get the resolved network name."""
        return self._network.name

    def get_leases(self) -> list[NetworkLeaseItem]:
        """Get all IP leases for this network.

        Returns:
            List of NetworkLeaseItem objects for the network.
        """
        db_leases = self._lease_repo.list_all(self._network.id)
        return [
            NetworkLeaseItem(
                network_id=lease.network_id,
                ipv4=lease.ipv4,
                vm_id=lease.vm_id,
                id=lease.id,
                leased_at=lease.leased_at,
                expires_at=lease.expires_at,
            )
            for lease in db_leases
        ]

    def get(self, ip: str) -> NetworkLeaseItem | None:
        """Get lease for a specific IP address.

        Args:
            ip: IP address to look up.

        Returns:
            NetworkLeaseItem if found, None otherwise.
        """
        lease = self._lease_repo.get(self._network.id, ip)
        if lease is None:
            return None
        return NetworkLeaseItem(
            network_id=lease.network_id,
            ipv4=lease.ipv4,
            vm_id=lease.vm_id,
            id=lease.id,
            leased_at=lease.leased_at,
            expires_at=lease.expires_at,
        )

    def get_by_vm_id(self, vm_id: str) -> list[NetworkLeaseItem]:
        """Get all leases for a specific VM on this network.

        Args:
            vm_id: VM ID to look up.

        Returns:
            List of NetworkLeaseItem objects for the VM.
        """
        db_leases = self._lease_repo.list_by_vm(self._network.id, vm_id)
        return [
            NetworkLeaseItem(
                network_id=lease.network_id,
                ipv4=lease.ipv4,
                vm_id=lease.vm_id,
                id=lease.id,
                leased_at=lease.leased_at,
                expires_at=lease.expires_at,
            )
            for lease in db_leases
        ]

    def is_available(self, ip: str) -> bool:
        """Check if an IP address is available (not leased).

        Queries the database directly to check availability.

        Args:
            ip: IP address to verify availability.

        Returns:
            True if the IP is not currently leased, False otherwise.
        """
        return self._lease_repo.get(self._network.id, ip) is None

    def lease(self, vm_id: str) -> str:
        """Allocate the next available IP from this network's subnet.

        Registers the lease in database.

        Args:
            vm_id: ID of the VM requesting the IP.
            config: Network configuration.

        Returns:
            The allocated IP address string.

        Raises:
            NetworkError: If no IPs available.
        """
        leases = self.get_leases()
        used_ips = {lease.ipv4 for lease in leases}
        allocated_ip = NetworkUtils.allocate_next_ip(
            list(used_ips),
            self._network.subnet,
            self._network.ipv4_gateway,
        )
        self._lease_repo.acquire(self._network.id, allocated_ip, vm_id)
        return allocated_ip

    def lease_specific(self, ip: str, vm_id: str) -> str:
        """Allocate a specific IP address from this network's subnet.

        Validates that the IP is in the subnet, not already leased, and not the gateway.

        Args:
            ip: The specific IP address to allocate.
            vm_id: ID of the VM requesting the IP.

        Returns:
            The allocated IP address string.

        Raises:
            NetworkError: If IP is not in subnet, is the gateway, or is already leased.
        """
        network = ipaddress.IPv4Network(self._network.subnet, strict=False)
        try:
            ip_obj = ipaddress.IPv4Address(ip)
            if ip_obj not in network:
                raise NetworkError(
                    f"IP {ip} is not in subnet {self._network.subnet}"
                )
        except ValueError as exc:
            raise NetworkError(f"Invalid IP address: {ip}") from exc

        if ip == self._network.ipv4_gateway:
            raise NetworkError(
                f"IP {ip} is the network gateway and cannot be allocated"
            )

        if not self.is_available(ip):
            raise NetworkError(f"IP {ip} is already leased")

        self._lease_repo.acquire(self._network.id, ip, vm_id)
        return ip

    def release(self, vm_id: str) -> None:
        """Release all leases for a VM from this network.

        Args:
            vm_id: ID of the VM whose leases should be released.
        """
        self._lease_repo.release_by_vm(vm_id)


__all__ = [
    "LeaseService",
]

File: internal/core/network/lease_service.go
package network

import (
	"context"
	"fmt"

	"mvmctl/internal/infra/errs"
	"mvmctl/internal/infra/model"
)

// LeaseService manages IP leases for a specific network.
// Matches src/mvmctl/core/network/_lease_service.py: LeaseService
type LeaseService struct {
	leaseRepo LeaseRepository
	net       *model.Network
}

// NewLeaseService creates a LeaseService from a *model.Network or string identifier.
// Matches Python: LeaseService(entity: str | NetworkItem, repo)
// Python's __init__ accepts both str and NetworkItem — if str, it resolves via
// NetworkResolver() which uses the default database. Go requires an explicit
// Repository for string resolution, and returns an error if networkRepo
// is nil when it's needed.
func NewLeaseService(ctx context.Context, entity any, leaseRepo LeaseRepository, networkRepo Repository) (*LeaseService, error) {
	switch e := entity.(type) {
	case *model.Network:
		return &LeaseService{
			leaseRepo: leaseRepo,
			net:       e,
		}, nil
	case string:
		if networkRepo == nil {
			return nil, errs.Wrap(errs.CodeNetworkNotFound,
				fmt.Errorf("cannot resolve network entity %q: no network repository provided", e))
		}
		resolver := NewResolver(networkRepo, nil)
		net, err := resolver.Resolve(ctx, e)
		if err != nil {
			return nil, err
		}
		return &LeaseService{
			leaseRepo: leaseRepo,
			net:       net,
		}, nil
	default:
		return nil, fmt.Errorf("expected *model.Network or string, got %T", entity)
	}
}

func (s *LeaseService) NetworkID() string {
	return s.net.ID
}

func (s *LeaseService) NetworkName() string {
	return s.net.Name
}

// GetLeases returns all IP leases for this network.
// Matches Python: returns new model.NetworkLeaseItem copies from db leases.
func (s *LeaseService) GetLeases(ctx context.Context) ([]*model.NetworkLeaseItem, error) {
	dbLeases, err := s.leaseRepo.ListAll(ctx, s.net.ID)
	if err != nil {
		return nil, err
	}
	result := make([]*model.NetworkLeaseItem, len(dbLeases))
	for i, l := range dbLeases {
		result[i] = &model.NetworkLeaseItem{
			NetworkID: l.NetworkID,
			IPv4:      l.IPv4,
			VMID:      l.VMID,
			ID:        l.ID,
			LeasedAt:  l.LeasedAt,
			ExpiresAt: l.ExpiresAt,
		}
	}
	return result, nil
}

// Get returns lease for a specific IP address.
// Matches Python: returns model.NetworkLeaseItem copy or nil.
func (s *LeaseService) Get(ctx context.Context, ip string) (*model.NetworkLeaseItem, error) {
	lease, err := s.leaseRepo.Get(ctx, s.net.ID, ip)
	if err != nil {
		return nil, err
	}
	if lease == nil {
		return nil, nil
	}
	return &model.NetworkLeaseItem{
		NetworkID: lease.NetworkID,
		IPv4:      lease.IPv4,
		VMID:      lease.VMID,
		ID:        lease.ID,
		LeasedAt:  lease.LeasedAt,
		ExpiresAt: lease.ExpiresAt,
	}, nil
}

// GetByVMID returns all leases for a specific VM on this network.
// Matches Python's get_by_vm_id.
func (s *LeaseService) GetByVMID(ctx context.Context, vmID string) ([]*model.NetworkLeaseItem, error) {
	dbLeases, err := s.leaseRepo.ListByVM(ctx, s.net.ID, vmID)
	if err != nil {
		return nil, err
	}
	result := make([]*model.NetworkLeaseItem, len(dbLeases))
	for i, l := range dbLeases {
		result[i] = &model.NetworkLeaseItem{
			NetworkID: l.NetworkID,
			IPv4:      l.IPv4,
			VMID:      l.VMID,
			ID:        l.ID,
			LeasedAt:  l.LeasedAt,
			ExpiresAt: l.ExpiresAt,
		}
	}
	return result, nil
}

// IsAvailable checks if an IP address is available (not leased).
// Matches Python's is_available.
func (s *LeaseService) IsAvailable(ctx context.Context, ip string) (bool, error) {
	lease, err := s.leaseRepo.Get(ctx, s.net.ID, ip)
	if err != nil {
		return false, err
	}
	return lease == nil, nil
}

// Lease allocates the next available IP from this network's subnet.
// Matches Python's lease() with max_retries=10 and IntegrityError handling.
func (s *LeaseService) Lease(ctx context.Context, vmID string) (string, error) {
	maxRetries := 10
	var lastError error

	for attempt := 0; attempt < maxRetries; attempt++ {
		leases, err := s.GetLeases(ctx)
		if err != nil {
			return "", err
		}
		usedIPs := make([]string, len(leases))
		for i, l := range leases {
			usedIPs[i] = l.IPv4
		}

		allocatedIP, err := AllocateNextIP(usedIPs, s.net.Subnet, s.net.IPv4Gateway)
		if err != nil {
			return "", errs.Wrap(errs.CodeNetworkLeaseExhausted, err)
		}

		vmIDCopy := vmID
		lease, err := s.leaseRepo.Acquire(ctx, s.net.ID, allocatedIP, &vmIDCopy)
		if err != nil {
			return "", err
		}
		if lease != nil {
			return allocatedIP, nil
		}

		// IP was already taken (INSERT OR IGNORE returned 0 rows affected) — retry next candidate.
		lastError = fmt.Errorf("IP %s already allocated", allocatedIP)
		continue
	}

	if lastError != nil {
		return "", errs.NetworkError(
			fmt.Sprintf("Failed to allocate IP after %d attempts", maxRetries))
	}
	return "", errs.NetworkError(
		fmt.Sprintf("Failed to allocate IP after %d attempts", maxRetries))
}

// LeaseSpecific allocates a specific IP address from this network's subnet.
// Matches Python's lease_specific.
func (s *LeaseService) LeaseSpecific(ctx context.Context, ip, vmID string) (string, error) {
	available, err := s.IsAvailable(ctx, ip)
	if err != nil {
		return "", err
	}
	if !available {
		return "", errs.NetworkError(fmt.Sprintf("IP %s is already leased", ip))
	}

	vmIDCopy := vmID
	_, err = s.leaseRepo.Acquire(ctx, s.net.ID, ip, &vmIDCopy)
	if err != nil {
		return "", err
	}
	return ip, nil
}

// Release releases all leases for a VM from this network.
// Matches Python's release.
func (s *LeaseService) Release(ctx context.Context, vmID string) error {
	return s.leaseRepo.ReleaseByVM(ctx, vmID)
}
